### ISSUE TITLE
JavaScript: default parameter causes type to be incorrectly inferred 

### ISSUE THREAD

**Issue by mjbvz** [Microsoft/TypeScript#28210]
From 

**TypeScript Version:**  3.2.0-dev.20181027

**Search Terms:**

**Code**
For the javascript:

```js
var SomeObject = {
    SomeFunction: function (SomeInput) {
        return SomeInput * 2;
    },
    SomeOtherFunction: function (SomeOtherInput, SomeOptionalInput = SomeObject.SomeFunction(2)) {
        return SomeOtherInput * SomeOptionalInput;
    }
};

SomeObject
```

Hover over `SomeObject`

**Bug:**
Type is `any`. If the default parameter is removed then the type is correct

**Comment by weswigham** [comment]
It looks like something causes us to pull on the signature type early which, with the initializer, depends on the object type itself and causes it to get marked as circular, specifically with `noImplicitAny` on, we get the
```ts
'SomeObject' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.
```
error. Thing is, if you wrote the same thing in a more namespace-y way:
```ts
namespace SomeObject {
    export function SomeFunction(SomeInput) {
        return SomeInput * 2;
    }
    export function SomeOtherFunction(SomeOtherInput, SomeOptionalInput = SomeObject.SomeFunction(2)) {
        return SomeOtherInput * SomeOptionalInput;
    }
}
```
we do waaaaaaaay better, even though the two are pretty much equivalent in this case. I think that _technically_ this is working as intended; but for the sake of our JS users, we could do better here.

**Comment by bcherny** [comment]
One more example of this:

```ts
type Currency = {
  unit: 'EUR' | 'GBP' | 'JPY' | 'USD'
  value: number
}

let Currency = { // TypeError 1
  DEFAULT: 'USD',
  from(value: number, unit = Currency.DEFAULT): Currency { // TypeError 2
    return {unit, value}
  }
}
```

It would be nice if this typechecked, since when you run it, it works as expected. Would this be a lot of work to improve?

**Comment by mjbvz** [comment]
@weswigham This looks fixed to me with TS 3.7.5. Can you please confirm

**Comment by weswigham** [comment]
I still see the behavior described in the OP in the playground, even on the [3.8 beta](http://www.typescriptlang.org/v2/en/play?strictNullChecks=true&ts=3.8.0-beta&useJavaScript=true#code/G4QwTgBAyg9gtgUwPICMBWCDGAXCBeCAbwCgIzp4EAxAVwDscBLGOgLggDN6mWIAKWIgCSdAA41sASiKlycsAmw0wdCsLESIAKggAmANyyyAXwA0RtcmwALBGFoNszNp25PeAykht2R47KaWSKLudCAANn6aBILI6FjYAHSxDjx0fLqS0iRy8orKqrHetmBRuDpFIc4RZYZyxsTGhsRF8ThAA).

**Comment by mjbvz** [comment]
Yeah I swear this was working but now it's back to `any` again when I tested it again. Would any jsconfig setting effect this?

**Comment by weswigham** [comment]
Errr.... `noImplicitAny`? `noImplicitThis`?
